pkg/pipeline/run.go

```diff
@@ -23,8 +23,81 @@ type terraformedInput struct {
 	ParametersTypeName string
 }
 
-// Run runs the Upjet code generation pipelines.
-func Run(pc *config.Provider, rootDir string) { //nolint:gocyclo
+// Run runs the Upjet code generation pipelines, using the given provider
+// configurations for cluster scoped and namespaced resources. The provider
+// configuration for namespaced resources is optional, if it isn't provided then
+// this function will only generate cluster scoped resources.
+func Run(pcCluster, pcNamespace *config.Provider, rootDir string) {
+	var groups []string
+	if pcNamespace == nil {
+		// namespaced resource generation is not enabled, generate only cluster scoped resources
+		cluster := &PipelineRunner{
+			DirAPIs:        filepath.Join(rootDir, "apis"),
+			DirControllers: filepath.Join(rootDir, "internal", "controller"),
+			DirExamples:    filepath.Join(rootDir, "examples-generated"),
+			DirHack:        filepath.Join(rootDir, "hack"),
+
+			ModulePathAPIs:        filepath.Join(pcCluster.ModulePath, "apis"),
+			ModulePathControllers: filepath.Join(pcCluster.ModulePath, "internal", "controller"),
+
+			Scope: "Cluster",
+		}
+
+		groups = cluster.Run(pcCluster)
+	} else {
```

Comment:
Instead of this pattern where most of this function is inside an `if/else` clause, I'd prefer a flow like:

```go
func Run(...) {
	// Handle the case where there's no namespaced config inside the smaller `if` block
	if pcNamespace == nil {
		cluster := ...
		if pcCluster.MainTemplate != "" { ... }
		return
	}

	// Handle the case where there's both with less nesting
	cluster := ...
	namespaced := ...
	if pcCluster.MainTemplate != "" { ... }
	return
}
```

Reply:
pushed https://github.com/crossplane/upjet/pull/480/commits/dc3707c290956fa9e04993c61a30bde00d67055e for this!